File: csharp/swift-scheduling/SwiftScheduling.cs
using System.Text.RegularExpressions;

public static class SwiftScheduling
{
    private static int MonthToQuarter(int month)
    {
        switch(month)
        {
            case 1:
            case 2:
            case 3:
                return 1;
            case 4:
            case 5:
            case 6:
                return 2;
            case 7:
            case 8:
            case 9:
                return 3;
            case 10:
            case 11:
            case 12:
                return 4;
            default:
                throw new Exception("Invalid Month only 1-12 are accepted.");
        }
    }

    private static DateTime LastDayOfQuarter(int quarter, int year)
    {
        if (quarter >= 1 && quarter <= 4)
        {
            DateTime temp;

            // April starts Q2, July Q3, October Q4, January Q1 next year.
            int month = (quarter * 3) + 1;
            int targetYear = year;
            if (month > 12)
            {
                month = 1;
                targetYear++;
            }
            temp = new DateTime(targetYear, month, 1);
            // subtract a day brings you to last day of desired quarter.
            temp = temp.AddDays(-1);
            return temp;
        }
        throw new Exception("Quarter out of bounds 1-4 only.");
    }

    public static DateTime DeliveryDate(DateTime meetingStart, string description) 
    {
        Regex quarter = new Regex(@"Q[1-4]", RegexOptions.IgnoreCase);
        Regex month = new Regex(@"\d+M", RegexOptions.IgnoreCase);
        DayOfWeek[] earlyWeek = {DayOfWeek.Monday, DayOfWeek.Tuesday, DayOfWeek.Wednesday};
        DayOfWeek[] weekEnd = {DayOfWeek.Saturday, DayOfWeek.Sunday};

        if (description == "NOW")
        {
            return meetingStart + new TimeSpan(2, 0, 0);
        }
        else if (description == "ASAP")
        {
            if (meetingStart.Hour < 13)
            {
                // Today at 17:00
                return new DateTime(meetingStart.Year
                                   , meetingStart.Month
                                   , meetingStart.Day
                                   , 17, 0, 0, 0);
            }
            else
            {
                // Tomorrow as 13:00
                return new DateTime(meetingStart.Year, 
                                    meetingStart.Month,
                                   meetingStart.Day
                                   , 13, 0, 0, 0).AddDays(1);
            }            
        }
        else if (description == "EOW")
        {
            if (earlyWeek.Contains<DayOfWeek>(meetingStart.DayOfWeek))
            {
                DateTime endOfWeek = new DateTime(meetingStart.Year, meetingStart.Month, meetingStart.Day, 17, 0, 0);
                while (endOfWeek.DayOfWeek != DayOfWeek.Friday)
                    endOfWeek = endOfWeek.AddDays(1);
                return endOfWeek;
            }
            else
            {
                DateTime endOfWeek = new DateTime(meetingStart.Year, meetingStart.Month, meetingStart.Day, 20, 0, 0);
                while (endOfWeek.DayOfWeek != DayOfWeek.Sunday)
                    endOfWeek = endOfWeek.AddDays(1);
                return endOfWeek;
            }
        }

        Match m = month.Match(description);
        if (m.Success)
        {
            bool success = int.TryParse(m.Value.ToUpperInvariant().Replace("M", ""), out int selectedMonth);
            if (!success) 
                throw new Exception("Invalid month format.");
            int year = meetingStart.Year;
            if (meetingStart.Month >= selectedMonth)
                year++;

            DateTime nthMonth = new DateTime(year, selectedMonth, 1, 8, 0, 0);
            while (weekEnd.Contains<DayOfWeek>(nthMonth.DayOfWeek))
                nthMonth = nthMonth.AddDays(1);
            return nthMonth;
        }

        Match q = quarter.Match(description);
        if (q.Success)
        {
            bool success = int.TryParse(q.Value.ToUpperInvariant().Replace("Q", ""), out int selectedQuarter);
            if (!success) 
                throw new Exception("Invalid quarter format.");
            int year = meetingStart.Year;
            if (MonthToQuarter(meetingStart.Month) > selectedQuarter)
                year++;

            DateTime nthQuarter = LastDayOfQuarter(selectedQuarter, year);
            while (weekEnd.Contains<DayOfWeek>(nthQuarter.DayOfWeek))
                nthQuarter = nthQuarter.AddDays(-1);
            return nthQuarter.AddHours(8);
        }

        
        return meetingStart;
    }
}

File: go/swift-scheduling/swift_scheduling.go
// Swift Scheduling Exercise
package swiftscheduling

import (
	"slices"
	"time"
)

var QUARTERS [][]time.Month = [][]time.Month{{time.January, time.February, time.March},
	{time.April, time.May, time.June},
	{time.July, time.August, time.September},
	{time.October, time.November, time.December},
}

var MONTHS []time.Month = []time.Month{
	time.January, time.February, time.March,
	time.April, time.May, time.June,
	time.July, time.August, time.September,
	time.October, time.November, time.December,
}

// For a given month, return the quarter (1-4)
// @param Month: The month to check for
// @returns: The quarter, and a boolean indicating an error occured
func GetQuarter(Month time.Month) (int, bool) {
	for i := 0; i < len(QUARTERS); i++ {
		if slices.Contains(QUARTERS[i], Month) {
			return i + 1, true
		}
	}
	return 0, false
}

// Calculate the actual date from a current date and delivery modifier value
// NOW - In two hours
// ASAP - Today at 5pm or Tomorrow at 1pm if current time is after 1pm
// EOW - End of Week for Monday, Tuesday, Wednesday Friday at 5pm, for the rest of the work week Sunday at 8pm
// [digit]M - 8:00 on first workday of the month of this year if before the month and next year if this month or later - digit is 1-12
// Q[digit] - 8:00 on the last workday of the last month of the given quarter (next year if quarter passed already this year) digit is 1-4
// @param start: The date the delivery time period was stated
// @param delivery: The delivery string NOW, ASAP, EOW, 2M, Q3, etc.
// @returns: Empty string on error otherwise the desired date formatted in 2026-01-02T15:04:05 format.
func DeliveryDate(start, delivery string) string {
	const FORMAT_STRING = "2006-01-02T15:04:05"

	StartDate, err := time.Parse(FORMAT_STRING, start)
	var ret time.Time
	var ok bool = true

	if err != nil {
		return ""
	}
	switch {
	case delivery == "NOW":
		ret = StartDate.Add(time.Hour * 2)
	case delivery == "ASAP":
		if StartDate.Hour() < 13 {
			ret = time.Date(StartDate.Year(), StartDate.Month(), StartDate.Day(), 17, 0, 0, 0, time.Local)
		} else {
			ret = time.Date(StartDate.Year(), StartDate.Month(), StartDate.Day(), 13, 0, 0, 0, time.Local)
			ret = ret.AddDate(0, 0, 1)
		}
	case delivery == "EOW":
		if StartDate.Weekday() == time.Monday || StartDate.Weekday() == time.Tuesday || StartDate.Weekday() == time.Wednesday {
			ret = time.Date(StartDate.Year(), StartDate.Month(), StartDate.Day(), 17, 0, 0, 0, time.Local)
			for ret.Weekday() != time.Friday {
				ret = ret.AddDate(0, 0, 1)
			}
		} else if StartDate.Weekday() == time.Thursday || StartDate.Weekday() == time.Friday {
			ret = time.Date(StartDate.Year(), StartDate.Month(), StartDate.Day(), 20, 0, 0, 0, time.Local)
			for ret.Weekday() != time.Sunday {
				ret = ret.AddDate(0, 0, 1)
			}
		} else {
			ok = false
		}
	case delivery[0] == 'Q':
		if delivery[1] == '1' || delivery[1] == '2' || delivery[1] == '3' || delivery[1] == '4' {
			var Quarter int = int(delivery[1]) - int('0')
			CurrentQuarter, ok := GetQuarter(StartDate.Month())
			if !ok {
				return ""
			}
			lastMonth := QUARTERS[Quarter-1][2]
			if CurrentQuarter <= Quarter {
				ret = time.Date(StartDate.Year(), lastMonth, 1, 8, 0, 0, 0, time.Local)
			} else {
				ret = time.Date(StartDate.Year()+1, lastMonth, 1, 8, 0, 0, 0, time.Local)
			}
			ret = ret.AddDate(0, 1, 0)  // Month forward
			ret = ret.AddDate(0, 0, -1) // Subtract a day for last day of month
			for ret.Weekday() == time.Saturday || ret.Weekday() == time.Sunday {
				ret = ret.AddDate(0, 0, -1) // Move backwards from weekend
			}
		} else {
			ok = false
		}
	case delivery[0] >= '0' && delivery[0] <= '9':
		MonthNum := 0
		var i int
		for i = 0; delivery[i] >= '0' && delivery[i] <= '9'; i++ {
			MonthNum *= 10
			MonthNum += int(delivery[i]) - int('0')
		}
		if delivery[i] == 'M' {
			Month := MONTHS[MonthNum-1]
			if StartDate.Month() < Month {
				ret = time.Date(StartDate.Year(), Month, 1, 8, 0, 0, 0, time.Local)
			} else {
				ret = time.Date(StartDate.Year()+1, Month, 1, 8, 0, 0, 0, time.Local)
			}
			for ret.Weekday() == time.Saturday || ret.Weekday() == time.Sunday {
				ret = ret.AddDate(0, 0, 1) // Move forward from weekend
			}
		} else {
			ok = false
		}
	}
	if !ok {
		return ""
	}
	return ret.Format(FORMAT_STRING)
}
